cpp/src/transform/arrow_to_cudf.cu

```diff
@@ -0,0 +1,263 @@
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+
+namespace detail {
+
+data_type arrow_to_cudf_type(arrow::Type::type arrow_type)
+{
+  switch (arrow_type) {
+    case arrow::Type::type::BOOL: return data_type(type_id::BOOL8);
+    case arrow::Type::type::INT8: return data_type(type_id::INT8);
+    case arrow::Type::type::INT16: return data_type(type_id::INT16);
+    case arrow::Type::type::INT32: return data_type(type_id::INT32);
+    case arrow::Type::type::INT64: return data_type(type_id::INT64);
+    case arrow::Type::type::UINT8: return data_type(type_id::UINT8);
+    case arrow::Type::type::UINT16: return data_type(type_id::UINT16);
+    case arrow::Type::type::UINT32: return data_type(type_id::UINT32);
+    case arrow::Type::type::UINT64: return data_type(type_id::UINT64);
+    case arrow::Type::type::FLOAT: return data_type(type_id::FLOAT32);
+    case arrow::Type::type::DOUBLE: return data_type(type_id::FLOAT64);
+    case arrow::Type::type::DATE32: return data_type(type_id::TIMESTAMP_DAYS);
+    case arrow::Type::type::DATE64: return data_type(type_id::TIMESTAMP_MILLISECONDS);
+    case arrow::Type::type::STRING: return data_type(type_id::STRING);
+    case arrow::Type::type::DICTIONARY: return data_type(type_id::DICTIONARY32);
+    case arrow::Type::type::LIST: return data_type(type_id::LIST);
+    default: CUDF_FAIL("Unsupported type_id conversion to cudf");
+  }
+}
+
+struct dispatch_to_cudf_col {
+  template <typename T>
+  std::enable_if_t<is_fixed_width<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
+    const std::shared_ptr<arrow::Array> arrow_col,
+    data_type type,
+    bool skip_mask,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto data_buffer = arrow_col->data()->buffers[1];
+    auto num_rows    = data_buffer->size() / sizeof(T);
+    auto has_nulls   = skip_mask ? false : arrow_col->null_bitmap_data() != nullptr;
+    auto col         = make_fixed_width_column(
+      type, num_rows, has_nulls ? mask_state::UNINITIALIZED : mask_state::UNALLOCATED, stream, mr);
+    auto mutable_column_view = col->mutable_view();
+    cudaMemcpy(mutable_column_view.data<void*>(),
+               data_buffer->data(),
+               data_buffer->size(),
+               cudaMemcpyHostToDevice);
+    if (has_nulls) {
+      cudaMemcpy(mutable_column_view.null_mask(),
+                 arrow_col->null_bitmap_data(),
+                 arrow_col->null_bitmap()->size(),
+                 cudaMemcpyHostToDevice);
+    }
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(arrow_col->offset()),
+                          static_cast<size_type>(arrow_col->length())));
+  }
+
+  template <typename T>
+  std::enable_if_t<is_compound<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
+    const std::shared_ptr<arrow::Array> arrow_col,
```

Comment:
```suggestion
    arrow::Array const& arrow_col,
```

Reply:
I don't think this will help as under compound types, I need to extract child class and making Array will omit subclass information.

Reply:
It's still being passed by reference, so you can still use it polymorphically.

Reply:
We get shared_ptr<arrow::Array> from arrow::Table, so dreferencing will give us just arrow::Array, and we will not be able to cast it to anything.

Reply:
I don't understand. In terms of polymorphism, there is nothing a `shared_ptr` can do that a `const&` cannot do.